### DISCUSSION TITLE
`@UseRequestContext()` breaks `@MessagePattern()` and `@EventPattern()` from @nestjs/microservices

### DISCUSSION THREAD

**Discussion by EvilCheetah** [mikro-orm/mikro-orm#4275] [Q&A] [answered]
**Describe the bug**
I have found out that adding `@UseRequestContext()` decorator to the methods covered with `@MessagePattern()` and `@EventPattern()` from `@nestjs/microservices` breaks behavior of the service. I attempted to investigate the problem on my own to avoid the need for this ticket, but after some examination and experimentation, I concluded that removing the `@UseRequestContext` allowed `@nestjs/microservices` decorators to function correctly as intended.
___
**Stack trace**
When `@UseRequestContext()` decorator applied with `@MessagePattern()`
On `gateway` service side:
```
[Nest] 55452  - 04/25/2023, 3:49:44 PM   ERROR [ExceptionsHandler] There is no matching message handler defined in the remote service.
```

When `@UseRequestContext()` decorator is applied with `@EventPattern()`
On `microservice`(rabbitmq-test) service side:
```
[Nest] 55531  - 04/25/2023, 3:49:19 PM   ERROR [Server] There is no matching event handler defined in the remote service. Event pattern: app.notify.event
```  
___
**To Reproduce**
Repository that reproduces the issue: [Repository](https://github.com/EvilCheetah/mikro-orm-context-error)

1. Install dependencies: `pnpm install`
2. Run core containers: `docker-compose up -d`
3. Migrate the database to the latest state:`mikro-orm migration:up`
4. Run an instance of **gateway**: `pnpm run start:dev gateway`
5. Run an instance of **rabbitmq-test**: `pnpm run start:dev rabbitmq-test`
6. Make `GET` and `POST` to ensure the work. Inspect `rabbitmq-test` terminal to see debug messages
7. Uncomment `@UseRequestContext()` decorators in `apps/rabbitmq-test/src/app.controller`
8. Repeat step (6), and inspect terminals of both `rabbitmq-test` and `gateway` services
___
**Expected behavior**
The original expectation was to utilize the **mikro-orm** `@EntityRepository<T>` seamlessly with `@nestjs/microservices` without requiring additional setup.
___
**Additional context**
This issue originally comes from [Discussion #4256](https://github.com/mikro-orm/mikro-orm/discussions/4256).
At present, my objective is to develop an application that employs microservices with the aid of NestJS, utilizing RabbitMQ as the message broker, and mikro orm as the core Object-Relational Mapping (ORM) tool.
___

**Versions**

| Dependency | Version |
| - | - |
| node | v18.12.0 |
| typescript | 5.0.4 |
| mikro-orm | 5.7.2 |
| your-driver | postgresql |

**Comment by B4nan** [comment, marked as answer]
The decorator is not doing much, it just wraps the method. Those issues are usually about the order of decorators. 

See the bottom of this section:

https://mikro-orm.io/docs/usage-with-nestjs#request-scoped-handlers-in-queues

**Reply by EvilCheetah** [reply]
Bro, you are amazing! Thanks for pointing this out to me, I really appreciate it! I should of better read the documentation.
___
However, I still have a question, is there a way to avoid decorating all methods with `@UseRequestContext()` that use EntityRepository in any way? Maybe, registering a middleware, or specifying a scope on `@Controller()` or `@Injectable()`?

**Reply by B4nan** [reply]
That is more of a nestjs question, maybe try asking there, I am not using nestjs myself, apart from maintaining some example apps. If I knew a better way, I would have told you in the first place :)

For reading stuff, you could also try this:

https://mikro-orm.io/docs/entity-manager#disabling-identity-map-and-change-set-tracking

**Reply by EvilCheetah** [reply]
I will ask the community on how to resolve the issue. I will keep you up-to-date about it.
___
After following the directions you provided, I have noticed that I must inject `MikroORM` into the service in parallel to `EntityRepository`. Should it be a right behavior though? Feels like adding more boilerplate.

**Reply by B4nan** [reply]
Yes, you need to have some context provider, the decorator itself does not know anything - it is a global function.

**Reply by simonbrunel** [reply]
> I have noticed that I must inject MikroORM into the service in parallel to EntityRepository.

This creates a strange pattern where some classes inject the `entityManager` and other ones the `orm`. I feel that `UseRequestContext` should have expected an instance of `EntityManager`, since it's what it needs to call `RequestContext.createAsync()`. Maybe we could improve the decorator to support objects containing the `entityManager: EntityManager` property and fallback to `orm` (maybe [something like this](https://gist.github.com/simonbrunel/d89c5f872ae467c537815f9e1cc2d74a))?

Another issue with `orm` is that *I think* we can't retrieve it from the `entityManager`, so we can't use the decorator callback:

```ts
// Should return an instance of MikroORM 
@UseRequestContext((type) => type.entityManager.???)
```

However, if the decorator expects the entity manager, cases that inject `MikroORM` can do this:

```ts
// Should return an instance of EntityManager 
@UseRequestContext((type) => type.orm.em)
```

**Reply by B4nan** [reply]
Yes, you should be injecting the ORM object - because that is where we store the global EM instance. Not sure what is so strange about it, it's fine to inject all of them in a single controller, depends on what you want to do. If you would want to work with the schema generator, you would also have to inject the ORM object, its not some dark pattern you should never use, its just fine.

But technically a fork from the global EM should work exactly the same as a fork from a fork, so yes, we could allow passing in the EM there too.

If you don't like how the decorator works, you can bring your own, it's not doing anything special, just wraps the method with `RequestContext.createAsync`.

**Reply by B4nan** [reply]
In the end, we can make it work with repositories too, as they now expose the EM getter.

**Reply by simonbrunel** [reply]
Thanks @B4nan for your reply!

> If you don't like how the decorator works, you can bring your own...

Yes, that's what [we did in our project](https://gist.github.com/simonbrunel/d89c5f872ae467c537815f9e1cc2d74a).

> Not sure what is so strange about it, it's fine to inject all of them in a single controller.

Ok, maybe *strange* is not the right word 😄 Most (if not all) of our services only rely on the entity manager and never have to use the orm directly. Meaning we always inject an instance of `EntityManager`, except for those services that actually need to use  `@UseRequestContext` (e.g. tasks scheduling services). And I noticed that in these services, the `EntityManager` is not injected but instead `this.orm.em` is used, which create some inconsistencies with the rest of the code that uses `this.entityManager`.

I'm a bit confused about the difference between the injected `EntityManager`, compared using `this.orm.em` from the injected `MikroORM` and I'm not sure what is a best practice.

> ...a fork from the global EM should work exactly the same as a fork from a fork...

Good you mentioned that, thanks!

**Reply by EvilCheetah** [reply]
As a visual learner, I find it more helpful to observe the implementation of something and go through it step by step to comprehend how it functions, in addition to read documentation and make use of it. I have tried to follow NestJS examples and a real-world application using MikroORM, and I attempted to apply their best practices to my work, but with a slightly different approach - microservices. However, I was caught off guard when I encountered these issues.
___
Decorating all methods seemed confusing to me on all the methods that lead to the use of `EntityRepository`. As a solution, I created a specialized service that encapsulates the EntityRepository logic and decorated all the methods with `@UseRequestContext()`.
It looks like this, which basically decorates all methods using `@UseRequestContext()`:
```ts
@Injectable()
export class RegionRepository
{
    constructor(
        private readonly orm: MikroORM,

        @InjectRepository(Region)
        private readonly regionRepository: EntityRepository<Region>
    ) {}


    @UseRequestContext()
    async create(createRegionDTO: CreateRegionDTO): Promise<Region>
    {
        const region = this.regionRepository.create(createRegionDTO);

        await this.regionRepository.persistAndFlush(region);

        return region;
    }


    @UseRequestContext()
    find_all(): Promise<Region[]>
    {
        return this.regionRepository.findAll();
    }


    @UseRequestContext()
    async update({region_id, ...data}: UpdateRegionDTO): Promise<Region>
    {
        const region = await this.find_one(region_id);

        wrap(region).assign(data);
        await this.regionRepository.persistAndFlush(region);

        return region;
    }


    @UseRequestContext()
    async remove(region_id: number): Promise<void>
    {
        const region = await this.find_one(region_id);

        await this.regionRepository.removeAndFlush(region);
    }
}
```

**Reply by EvilCheetah** [reply]
I have tried to ask people on the NestJS Discord server right after your message, but nobody seems to be answering it.
___
I have tried to read the files in `@mikro-orm/nestjs` and it seems you are registering the http-paths automatically using `configure()`-method in `mikro-orm-core.module.ts` file, which uses `forRoutesPath()` helper method.
I understand that adding `@nestjs/microservices` decorators might be a temporary solution, due to the possibility of the existence of other packages in the present and future, offering different approaches to handle the situation. Is it possible to incorporate an abstraction layer that can automatically register the methods or provide the user with the ability to configure the `mikro-orm module` to indicate the required decorators or methods?
___
Don't get me wrong, I love MikroORM, and, in my opinion, it is the best solution on the market.

**Reply by B4nan** [reply]
Decorating repository methods is almost always wrong - this decorator will create new context for each call, I doubt you want that. The context should be a logical unit, let it be a request handler, or some job/message handler - but not a single repository call, you want to share one context for multiple calls under one job handler. 

E.g. your `remove` method - it calls `find_one`, another decorated method - those two calls will have their own context. v6 adds a new `@EnsureRequestContext` decorator that would be better suited for this, but still, it's generally not a good idea, I'd guess you will often need a shared context for multiple repository methods, so you will still need the decorator on higher level (some controller maybe).

> Is it possible to incorporate an abstraction layer that can automatically register the methods or provide the user with the ability to configure the mikro-orm module to indicate the required decorators or methods?

Not sure what you mean by that, can you be more specific, maybe provide some examples?

I was looking at nestjs docs about this, looks like they do support request-scoped services, so maybe that could be used? 

https://docs.nestjs.com/microservices/basics#scopes

The nestjs adapter can be used to define repositories in the nestjs request scope:

https://github.com/mikro-orm/nestjs#using-nestjs-injection-scopes-for-request-context
